File: data/figures/chapter06/figure_6_6/plot_6_6.py
#!/usr/bin/env python3
"""Replot this thesis figure from the sibling CSV only."""
from pathlib import Path
import argparse
import csv
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np

HERE = Path(__file__).resolve().parent
DATA = HERE / "figure_6_6.csv"
parser = argparse.ArgumentParser(description="Replot thesis Figure 6.6 from its sibling CSV.")
parser.add_argument("--output", type=Path, help="Output image path; defaults beside this script.")
arguments = parser.parse_args()
OUTPUT = (arguments.output if arguments.output is not None else HERE / "plot_6_6.png").resolve()
OUTPUT.parent.mkdir(parents=True, exist_ok=True)

plt.rcParams.update({"font.family": "serif", "font.serif": ["Times New Roman", "DejaVu Serif"], "font.size": 9})

def rows_from(path):
    with path.open(newline="", encoding="utf-8-sig") as handle:
        return list(csv.DictReader(handle))

def number(row, key): return float(row[key])
def flag(value): return str(value).strip().lower() == "true"

rows=[r for r in rows_from(DATA) if flag(r["c_feasible_label"])]
keys=["x1_ar","x2_taper_ratio","x3_primary_member_ratio","x4_secondary_member_ratio","two_wing_mass_kg","y1_cditrim","y2_ctrim_nm"]
labels=["AR","lambda","r1","r2","m2W","CDi,trim","Ctrim"]
bounds={"x1_ar":(6,12),"x2_taper_ratio":(.2,.8),"x3_primary_member_ratio":(.05,.4),"x4_secondary_member_ratio":(.15,.5)}
ranges={k:(min(number(r,k) for r in rows),max(number(r,k) for r in rows)) for k in keys}; ranges.update(bounds)
def scaled(r): return [(number(r,k)-ranges[k][0])/(ranges[k][1]-ranges[k][0]) for k in keys]
fig,ax=plt.subplots(figsize=(7.2,4.9)); x=np.arange(len(keys))
for r in rows:
    if flag(r["final_pareto"]) and r["phase"]=="adaptive_phase": color,lw,alpha="#E66101",1.8,.92
    elif flag(r["final_pareto"]): color,lw,alpha="#222222",1.35,.78
    else: color,lw,alpha="#2878B5",.8,.28
    ax.plot(x,scaled(r),color=color,lw=lw,alpha=alpha)
for i in range(len(keys)): ax.plot([i,i],[0,1],color="#222222",lw=.65)
ax.set_xticks(x,labels); ax.set_yticks([]); ax.set_xlim(-.2,len(keys)-.8); ax.set_ylim(-.03,1.03)
ax.set_title("Parallel Coordinates of the 70 Feasible Evaluations\n"
             "Mass is diagnostic; the two rightmost axes are minimized objectives")
for spine in ax.spines.values(): spine.set_visible(False)
from matplotlib.lines import Line2D
ax.legend(handles=[Line2D([0],[0],color="#2878B5",label="Feasible, dominated"),Line2D([0],[0],color="#222222",label="Initial-design Pareto"),Line2D([0],[0],color="#E66101",label="Adaptive Pareto")],loc="lower center",bbox_to_anchor=(.5,-.16),ncol=3,fontsize=7)
fig.tight_layout(); fig.savefig(OUTPUT,dpi=200,bbox_inches="tight"); plt.close(fig); print(OUTPUT)

File: data/figures/chapter06/figure_6_6/plot_6_6.m
function outputFile = plot_6_6(outputFile)
scriptDir = fileparts(mfilename('fullpath'));
if nargin < 1 || isempty(outputFile)
    outputFile = fullfile(scriptDir, 'plot_6_6.png');
end
outputFile = char(outputFile);
outputDir = fileparts(outputFile);
if ~isempty(outputDir) && ~isfolder(outputDir), mkdir(outputDir); end
dataFile = fullfile(scriptDir, 'figure_6_6.csv');
T = readtable(dataFile, 'TextType', 'string', 'VariableNamingRule', 'preserve');
set(groot, 'defaultAxesFontName', 'Times New Roman'); set(groot, 'defaultTextFontName', 'Times New Roman');
T=T(truth(T.c_feasible_label),:); keys={'x1_ar','x2_taper_ratio','x3_primary_member_ratio','x4_secondary_member_ratio','two_wing_mass_kg','y1_cditrim','y2_ctrim_nm'}; labels={'AR','lambda','r1','r2','m2W','CDi,trim','Ctrim'}; X=zeros(height(T),numel(keys));
for j=1:numel(keys), X(:,j)=T.(keys{j}); end
bounds=[6 12;.2 .8;.05 .4;.15 .5];
for j=1:4, X(:,j)=(X(:,j)-bounds(j,1))/(bounds(j,2)-bounds(j,1)); end
for j=5:7, X(:,j)=(X(:,j)-min(X(:,j)))/(max(X(:,j))-min(X(:,j))); end
fig=figure('Color','w','Visible','off','Position',[100 100 950 620]); ax=axes(fig); hold(ax,'on');
for i=1:height(T)
 if truth(T.final_pareto(i)) && T.phase(i)=="adaptive_phase", c=[.90 .38 .01]; w=1.8; a=.92; elseif truth(T.final_pareto(i)), c=[.13 .13 .13]; w=1.35; a=.78; else, c=[.16 .47 .71]; w=.8; a=.28; end
 plot(ax,1:7,X(i,:),'Color',[c a],'LineWidth',w);
end
for j=1:7, plot(ax,[j j],[0 1],'Color',[.13 .13 .13],'LineWidth',.65); end
xlim(ax,[.8 6.2]); ylim(ax,[-.03 1.03]); set(ax,'XTick',1:7,'XTickLabel',labels,'YTick',[]); box(ax,'off'); title(ax,sprintf('Parallel Coordinates of the 70 Feasible Evaluations\nMass is diagnostic; the two rightmost axes are minimized objectives'));
h1=plot(ax,nan,nan,'Color',[.16 .47 .71],'LineWidth',1.2); h2=plot(ax,nan,nan,'Color',[.13 .13 .13],'LineWidth',1.5); h3=plot(ax,nan,nan,'Color',[.90 .38 .01],'LineWidth',2); legend(ax,[h1 h2 h3],{'Feasible, dominated','Initial-design Pareto','Adaptive Pareto'},'Location','southoutside','NumColumns',3);
exportgraphics(fig,outputFile,'Resolution',200); close(fig); disp(outputFile);
end
function v=truth(x), if islogical(x),v=x; elseif isnumeric(x),v=x~=0; else,v=strcmpi(string(x),'true'); end, end
